lib/limits/estimates.js

```diff
@@ -0,0 +1,115 @@
+// Functions for estimating the number of rows produced by nodes and other quantities
+
+'use strict';
+
+var async = require('async');
+var dot = require('dot');
+
+var estimatedCountTemplate = dot.template('EXPLAIN (FORMAT JSON) {{=it.sourceQuery}}');
+var exactCountTemplate = dot.template([
+        'SELECT count(*) AS result_rows',
+        'FROM ({{=it.sourceQuery}}) _cdb_analysis_source'
+    ].join('\n'));
+var countDistinctValuesTemplate = dot.template([
+    'SELECT COUNT(*) AS count_distict_values FROM (',
+    '  SELECT DISTINCT {{=it.column}}',
+    '  FROM (',
+    '    {{=it.source}}',
+    '  ) AS _cdb_analysis_query',
+    ') AS _cdb_analysis_count_distinct'
+    ].join('\n'));
+
+// Estimated number of rows in the result of a node.
+// A LimitsContext object must be provided.
+// This is an asynchronous method; an error object and the number of estimated rows will
+// be passed to the callback provided.
+// This uses PostgreSQL query planner and statistics; it could fail, returning a null count
+// if no stats are available or if a timeout occurs.
+function estimatedNumberOfRows(node, context, callback) {
+    var sql = estimatedCountTemplate({
+        sourceQuery: node.sql()
+    });
+    context.runSQL(sql, function(err, resultSet){
+        var estimated_rows = null;
+        if (!err) {
+            estimated_rows = resultSet.rows[0]['QUERY PLAN'][0].Plan['Plan Rows'];
+        }
+        return callback(err, estimated_rows);
+    });
+}
+
+// Get estimated number of rows of all the node inputs; passes an object to the callback:
+// { input_param: estimatedNumberOfRows, ... }
+function estimatedNumberOfInputRows(node, context, callback) {
+    // much easier with async 2.0!
+    // async.mapValues(node.nodes, function(node, param, done) {
+    //     var sql = estimatedCountTemplate({
+    //         sourceQuery: node.sql()
+    //     });
+    //     context.runSQL(sql, function(err, resultSet){
+    //         var estimated_rows = null;
+    //         if (!err) {
+    //             estimated_rows = resultSet.rows[0]['QUERY PLAN'][0].Plan['Plan Rows'];
+    //         }
+    //         return done(err, estimated_rows);
+    //     });
+    //
+    // }, callback);
+    var results = {};
+    async.forEachOf(node.nodes, function(node, param, done) {
```

Comment:
Yes, I was wondering how to handle this, do you know how could I get the input nodes and the parameter name they correspond to easily? I'm trying to decide between keeping the same functionality avoiding using node.nodes and changing the functions that estimate number of input rows to require always the specific input nodes to be checked...